plugins/processors/gpuattributes/processor.go

```diff
@@ -0,0 +1,176 @@
+	gpuNodeMetricPrefix      = "node_"
+)
+
+var podContainerMetricLabels = map[string]map[string]interface{}{
+	"ClusterName":  nil,
+	"FullPodName":  nil,
+	"PodName":      nil,
+	"InstanceId":   nil,
+	"InstanceType": nil,
+	"NodeName":     nil,
+	"Timestamp":    nil,
+	"Type":         nil,
+	"Version":      nil,
+	"Namespace":    nil,
+	"Sources":      nil,
+	"UUID":         nil,
+	"Service":      nil,
+	"GpuDevice":    nil,
+	"kubernetes":   nil,
+}
+
+var nodeMetricLabels = map[string]map[string]interface{}{
+	"ClusterName":  nil,
+	"InstanceId":   nil,
+	"InstanceType": nil,
+	"NodeName":     nil,
+	"Timestamp":    nil,
+	"Type":         nil,
+	"Version":      nil,
+	"kubernetes": {
+		"host": nil,
+	},
+}
+
+type gpuprocessor struct {
+	*Config
+	logger *zap.Logger
+}
+
+func newGpuProcessor(config *Config, logger *zap.Logger) *gpuprocessor {
+	d := &gpuprocessor{
+		Config: config,
+		logger: logger,
+	}
+	return d
+}
+
+func (d *gpuprocessor) processMetrics(_ context.Context, md pmetric.Metrics) (pmetric.Metrics, error) {
+
+	rms := md.ResourceMetrics()
+	for i := 0; i < rms.Len(); i++ {
+		rs := rms.At(i)
+		ilms := rs.ScopeMetrics()
+		for j := 0; j < ilms.Len(); j++ {
+			ils := ilms.At(j)
+			metrics := ils.Metrics()
+			for k := 0; k < metrics.Len(); k++ {
+				m := metrics.At(k)
+				d.processMetricAttributes(m)
+			}
+		}
+	}
+	return md, nil
+}
+
+func (d *gpuprocessor) processMetricAttributes(m pmetric.Metric) {
+	// only decorate GPU metrics
+	// another option is to separate GPU of its own pipeline to minimize extra processing of metrics
+	if !strings.Contains(m.Name(), gpuMetric) {
+		return
+	}
+
+	var labels map[string]map[string]interface{}
+	if strings.HasPrefix(m.Name(), gpuNodeMetricPrefix) {
+		labels = nodeMetricLabels
+	} else if strings.HasPrefix(m.Name(), gpuContainerMetricPrefix) {
+		labels = podContainerMetricLabels
+		labels["ContainerName"] = nil
+		labels["kubernetes"] = map[string]interface{}{
+			"container_name": nil,
+			"containerd":     nil,
+			"host":           nil,
+			"labels":         nil,
+			"pod_id":         nil,
+			"pod_name":       nil,
+			"pod_owners":     nil,
+			"namespace":      nil,
+		}
+	} else if strings.HasPrefix(m.Name(), gpuPodMetricPrefix) {
+		labels = podContainerMetricLabels
+		labels["kubernetes"] = map[string]interface{}{
+			"host":       nil,
+			"labels":     nil,
+			"pod_id":     nil,
+			"pod_name":   nil,
+			"pod_owners": nil,
+			"namespace":  nil,
+		}
+	}
+
+	var dps pmetric.NumberDataPointSlice
+	switch m.Type() {
+	case pmetric.MetricTypeGauge:
+		dps = m.Gauge().DataPoints()
+	case pmetric.MetricTypeSum:
+		dps = m.Sum().DataPoints()
+	default:
+		d.logger.Debug("Ignore unknown metric type", zap.String("type", m.Type().String()))
+	}
+
+	for i := 0; i < dps.Len(); i++ {
+		d.filterAttributes(dps.At(i).Attributes(), labels)
+	}
+}
+
+func (d *gpuprocessor) filterAttributes(attributes pcommon.Map, labels map[string]map[string]interface{}) {
```

Comment:
If I'm right, then another option, since you have the `decoratorConsumer` in the `awscontainerinsightsreceiver` now, would be to do the container -> pod/node duplication there instead of here.  Possibly nicer so that you can be sure the metrics coming out of `awscontainerinsightreceiver` are all fully processed already.

Reply:
Yeah, your understanding is correct. The current state of structure/design is the outcome of multiple iterations of changes. 
We could make `awscontainerinsightreceiver` more than a receiver to "fully" process metrics, but I would like to utilize what's already available as plugins without duplicating or reinventing them. My intention was to let a plugin serves its original purpose. Like, receiver scrapes/gets data (not 100% sure this is totally true with the GPU decorator in the picture now), processors (`metrictransformer` and `gpuattributes`) process them by adding attributes or duplicating metrics and exporter exports without massaging the data it gets from processors.